provider/packet/environ.go

```diff
@@ -0,0 +1,322 @@
+	"github.com/juju/juju/core/constraints"
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/environs"
+	environscloudspec "github.com/juju/juju/environs/cloudspec"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/environs/context"
+	"github.com/juju/juju/environs/instances"
+	envinstance "github.com/juju/juju/environs/instances"
+	"github.com/juju/juju/provider/common"
+	"github.com/juju/juju/storage"
+	"github.com/juju/schema"
+	"github.com/juju/utils/arch"
+	"github.com/juju/version"
+	"github.com/packethost/packngo"
+	"gopkg.in/juju/environschema.v1"
+)
+
+type environConfig struct {
+	config *config.Config
+	attrs  map[string]interface{}
+}
+
+type environ struct {
+	ecfgMutex    sync.Mutex
+	ecfg         *environConfig
+	name         string
+	cloud        environscloudspec.CloudSpec
+	packetClient *packngo.Client
+}
+
+var providerInstance environProvider
+
+func (e *environ) AdoptResources(ctx context.ProviderCallContext, controllerUUID string, fromVersion version.Number) error {
+	trace()
+	return nil
+}
+
+func (e *environ) Bootstrap(ctx environs.BootstrapContext, callCtx context.ProviderCallContext, args environs.BootstrapParams) (*environs.BootstrapResult, error) {
+	trace()
+	return common.Bootstrap(ctx, e, callCtx, args)
+}
+
+func (e *environ) AllInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag("juju-instance", "")
+}
+
+// if values tag and state are left empty it will return all instances
+func (e *environ) getPacketInstancesByTag(tag, state string) ([]instances.Instance, error) {
+	toReturn := []instances.Instance{}
+	opt := &packngo.ListOptions{Search: tag}
+
+	devices, _, err := e.packetClient.Devices.List(e.cloud.Credential.Attributes()["project-id"], opt)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, d := range devices {
+		if state != "" && d.State == state {
+			toReturn = append(toReturn, &packetDevice{e, &d})
+		} else if state == "" {
+			toReturn = append(toReturn, &packetDevice{e, &d})
+		}
+	}
+
+	return toReturn, nil
+}
+
+func (e *environ) AllRunningInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag("juju-instance", "active")
+}
+
+func (e *environ) Config() *config.Config {
+	e.ecfgMutex.Lock()
+	defer e.ecfgMutex.Unlock()
+	return e.ecfg.config
+}
+
+func (e *environ) ConstraintsValidator(ctx context.ProviderCallContext) (constraints.Validator, error) {
+	validator := constraints.NewValidator()
+	validator.RegisterUnsupported([]string{constraints.CpuPower, constraints.VirtType})
+	validator.RegisterConflicts([]string{constraints.InstanceType}, []string{constraints.Mem})
+	validator.RegisterVocabulary(constraints.Arch, []string{arch.AMD64, arch.ARM64, arch.I386, arch.PPC64EL})
+	return validator, nil
+}
+
+func (e *environ) ControllerInstances(ctx context.ProviderCallContext, controllerUUID string) ([]instance.Id, error) {
+	trace()
```

Comment:
I think that this function should be returning all the controller instances. Since this is just a PR draft I was thinking to get suggestions from you.

Reply:
Since the packet API seems to support tags, you can apply a `jujucontroller` tag when bootstrapping and then call your getPacketInstancesByTag helper to grab the controllers (note, you may have more than one controllers when running in HA mode)